File: JavaScript/src/devicedriver.js
const nanoTime = require('nano-time');

const INIT_ADDRESS = 0x00;

const PROGRAM_COMMAND = 0x40;
const READY_MASK = 0x02;
const READY_NO_ERROR = 0x00;

const TIMEOUT_THRESHOLD = 100000000;

const RESET_COMMAND = 0xFF;
const VPP_MASK = 0x20;
const INTERNAL_ERROR_MASK = 0x10;
const PROTECTED_BLOCK_ERROR_MASK = 0x08;

/**
 * This class is used by the operating system to interact with the hardware.
 */
class DeviceDriver {
    constructor(hardware) {
        this.hardware = hardware;
    }

    read(address) {
        return this.hardware.read(address);
    }

    write(address, data) {
        let start = nanoTime();
        this.hardware.write(INIT_ADDRESS, PROGRAM_COMMAND);
        this.hardware.write(address, data);
        let readyByte;
        while (((readyByte = this.read(INIT_ADDRESS)) & READY_MASK) == 0) {
            if (readyByte != READY_NO_ERROR) {
                this.hardware.write(INIT_ADDRESS, RESET_COMMAND);
                if ((readyByte & VPP_MASK) > 0) {
                    throw new VppException();
                }
                if ((readyByte & INTERNAL_ERROR_MASK) > 0) {
                    throw new InternalErrorException();
                }
                if ((readyByte & PROTECTED_BLOCK_ERROR_MASK) > 0) {
                    throw new ProtectedBlockException();
                }
            }
            if (nanoTime() - start > TIMEOUT_THRESHOLD) {
                throw new TimeoutException('Timeout when trying to read data from memory');
            }
        }
        const actual = this.read(address);
        if (data != actual) {
            throw ReadFailureException('Failed to read data from memory');
        }
    }
}

class VppException extends Error {
    constructor() {
        super('vpp exception');
    }
}

class InternalErrorException extends Error {
    constructor() {
        super('internal error');
    }
}

class ProtectedBlockException extends Error {
    constructor() {
        super('protected block');
    }
}

class TimeoutException extends Error {
    constructor(message) {
        super(message);
    }
}

class ReadFailureException extends Error {
    constructor(message) {
        super(message);
    }
}

module.exports = {
    DeviceDriver,
    VppException,
    InternalErrorException,
    ProtectedBlockException,
    TimeoutException,
    ReadFailureException
};

File: Java/src/main/java/devicedriver/DeviceDriver.java
package devicedriver;

/**
 * This class is used by the operating system to interact with the hardware 'FlashMemoryDevice'.
 */
public class DeviceDriver {

  private static final long INIT_ADDRESS = 0X0;
  private static final byte PROGRAM_COMMAND = 0X40;
  private static final byte READY_MASK = 0x02;
  private static final byte READY_NO_ERROR = 0x00;
  private static final long TIMEOUT_THRESHOLD = 100_000_000;
  private static final byte RESET_COMMAND = (byte) 0xFF;
  private static final byte VPP_MASK = 0x20;
  private static final byte INTERNAL_ERROR_MASK = 0x10;
  private static final byte PROTECTED_BLOCK_ERROR_MASK = 0X08;

  private final FlashMemoryDevice hardware;

  public DeviceDriver(FlashMemoryDevice hardware) {
    this.hardware = hardware;
  }

  public byte read(long address) {
    return hardware.read(address);
  }

  public void write(long address, byte data) {
    long start = System.nanoTime();
    hardware.write(INIT_ADDRESS, PROGRAM_COMMAND);
    hardware.write(address, data);
    byte readyByte;
    while (((readyByte = read(INIT_ADDRESS)) & READY_MASK) == 0) {
      if (readyByte != READY_NO_ERROR) {
        hardware.write(INIT_ADDRESS, RESET_COMMAND);
        if ((readyByte & VPP_MASK) > 0) {
          throw new VppException();
        }
        if ((readyByte & INTERNAL_ERROR_MASK) > 0) {
          throw new InternalErrorException();
        }
        if ((readyByte & PROTECTED_BLOCK_ERROR_MASK) > 0) {
          throw new ProtectedBlockException();
        }
      }
      if (System.nanoTime() - start > TIMEOUT_THRESHOLD) {
        throw new TimeoutException("Timeout when trying to read data from memory");
      }
    }
    byte actual = read(address);
    if (data != actual) {
      throw new ReadFailureException("Failed to read data from memory");
    }
  }
}
